### PR TITLE
Feature/processor class

### PR DESCRIPTION
**Pull request by ChristiaensBert**
Added some additional features and improvements (and tested them)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -173,4 +173,6 @@ tags
 # ruff
 .ruff_cache
 
-wandb
\ No newline at end of file
+wandb
+
+src/controlnet_aux/main.py
```
Comment (pdoane) [inline .gitignore:178]: This is probably fine but a little unexpected?

File: setup.py [hunk 2]
```diff
@@ -177,6 +177,7 @@ def run(self):
     deps["torchvision"],
     deps["timm"],
     deps["scikit-image"],
+    deps["mediapipe"],
```
Comment (pdoane) [inline setup.py:180]: You can look through the history here and it was requested for mediapipe to not be an explicit dependency.  The detector works if it is installed, and the rest of the library works if it is not.

File: src/controlnet_aux/tests/test_processor.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
+"""Test the Processor class."""
```
Comment (pdoane) [inline src/controlnet_aux/tests/test_processor.py:1]: We overlapped here and I added initial unit tests in #49 but using pytest.  @patrickvonplaten - which approach would you prefer?

File: src/controlnet_aux/processor.py [hunk 4]
```diff
@@ -20,6 +20,7 @@
                             ZoeDetector,
                             MediapipeFaceDetector)
 
+CHOICES = "Choices for the processor are hed, midas, mlsd, openpose, pidinet, normalbae, lineart, lineart_coarse, lineart_anime, canny, content_shuffle, zoe, mediapipe_face"
```
Comment (pdoane) [inline src/controlnet_aux/processor.py:23]: This could be derived from MODELS.keys() as one less place to update.

**Comment by pdoane** [comment]
I suggest having the MODEL names be consistent with the A1111 extension, e.g:

    "canny"
    "depth_midas"
    "depth_zoe"
    "lineart_anime"
    "lineart_coarse"
    "lineart_realistic"
    "mediapipe_face"
    "mlsd"
    "normal_bae"
    "normal_midas"
    "openpose"
    "openpose_face"
    "openpose_faceonly"
    "openpose_full"
    "openpose_hand"
    "scribble_hed
    "scribble_pidinet"
    "shuffle"
    "softedge_hed"
    "softedge_hedsafe"
    "softedge_pidinet"
    "softedge_pidsafe"

File: .gitignore (hunk 1 above)
Comment (ChristiaensBert) [reply .gitignore:178]: Yes, it was from local testing

File: setup.py (hunk 2 above)
Comment (ChristiaensBert) [reply setup.py:180]: Okay, my bad!

**Comment by ChristiaensBert** [comment]
Nice, where did you find the naming for the AUTOMATIC111 extension?

File: src/controlnet_aux/tests/test_processor.py (hunk 3 above)
Comment (ChristiaensBert) [reply src/controlnet_aux/tests/test_processor.py:1]: I added a pytest script too

**Comment by ChristiaensBert** [comment]
I changed the naming and the other stuff. I tried to use the correct arguments for each setting, let me know if something is still off!

File: src/controlnet_aux/processor.py [hunk 5]
```diff
@@ -20,45 +20,64 @@
...
+    'depth_midas': {'resize': 512},
     'mlsd': {'resize': False},
-    'openpose': {'resize': False, 'hand_and_face': True},
-    'pidinet': {'resize': False, 'safe': True},
-    'normalbae': {'resize': False},
-    'lineart': {'resize': False, 'coarse': True},
+    'openpose': {'resize': False, 'hand_and_face': False},
+    'openpose_face': {'resize': False, 'hand_and_face': True},
+    'openpose_faceonly': {'resize': False, 'include_body': False, 'include_face': True},
+    'openpose_full': {'resize': False, 'hand_and_face': True},
```
Comment (pdoane) [inline src/controlnet_aux/processor.py:64]: "openpose_face" -> include_body=True, include_hand=False, include_face=True
"openpose_faceonly" -> include_body=False, include_hand=False, include_face=True
"openpose_full" -> include_body=True, include_hand=True, include_face=True
"openpose_hand" -> include_body=True, include_hand=True, include_face=False

**Comment by pdoane** [comment]
Thanks for the updates!  We should get this merged first and then I'll resolve the conflicts in my PR (which removes the `resize` parameter as all the detectors are consistent on behavior with it).

Maybe we only need one unit test implementation in this PR though?  The pytest version looks better to me, and is also what diffusers uses so is my preference.

File: src/controlnet_aux/processor.py
Comment (jinwonkim93) [inline src/controlnet_aux/processor.py:1]: Cool!. It would be nice to add SamDetector in the processors.

File: src/controlnet_aux/processor.py [hunk 6]
```diff
@@ -20,84 +20,117 @@
...
+            Processor: controlnet aux processor
         """
+        if processor_id not in MODELS:
+            raise ValueError(f"Processor {processor_id} not found. {CHOICES}")
+
         processor = MODELS[processor_id]['class']
 
+        # check if the proecssor is a checkpoint model
         if MODELS[processor_id]['checkpoint']:
             processor = processor.from_pretrained("lllyasviel/Annotators")
```
Comment (jinwonkim93) [inline src/controlnet_aux/processor.py:122]: If you add SamDetector, this might have conflict because SAM weight is not on lllyasviel/Annotators

**Review by jinwonkim93 — COMMENTED** [review]
Cool. can you also add SamDetector?

File: src/controlnet_aux/tests/test_processor.py (hunk 3 above)
Comment (patrickvonplaten) [reply src/controlnet_aux/tests/test_processor.py:1]: Think we can def add two test files here no? Would be very nice to have some tests :-)

Comment (patrickvonplaten) [reply src/controlnet_aux/tests/test_processor.py:1]: Think we can def add two test files here no? Would be very nice to have some tests :-)

File: src/controlnet_aux/processor.py [hunk 7]
```diff
@@ -20,84 +20,117 @@
...
-        Returns:
-            Processor: Processor object
+            processor_id (str): processor name, options are 'hed, midas, mlsd, openpose,
+                                pidinet, normalbae, lineart, lineart_coarse, lineart_anime,
+                                canny, content_shuffle, zoe, mediapipe_face
+            params (Optional[Dict]): parameters for the processor
         """
-        print(f"Loading {processor_id} processor")
+        LOGGER.info("Loading %s".format(processor_id))
         self.processor_id = processor_id
```
Comment (patrickvonplaten) [inline src/controlnet_aux/processor.py:93]:
```suggestion
        if processor_id not in MODEL_PARAMS:
            raise ValueError(f"{processor_id} is not a valid processor id. Please make sure to choose one of {', '.join(MODEL_PARAMS.keys())}".)
        self.processor_id = processor_id
```

**Review by patrickvonplaten — COMMENTED** [review]
Nice improvement, left a small suggestion. Apart from this, we can merge it I think :-)

File: src/controlnet_aux/processor.py (hunk 6 above)
Comment (ChristiaensBert) [reply src/controlnet_aux/processor.py:122]: I will add SAM in the next PR, currently it seems to need a local checkpoint
```
sam = SamDetector.from_pretrained("./weight_path")
```

File: src/controlnet_aux/processor.py (hunk 7 above)
Comment (ChristiaensBert) [reply src/controlnet_aux/processor.py:93]: I do the check in the `self.load_processor(self.processor_id)` but I can move it to here if you want

**Comment by ChristiaensBert** [comment]
I made the changes! Other features, like integrating SAM (with a local checkpoint?) can be done in a next PR? Other ideas are certainly welcome

File: src/controlnet_aux/processor.py (hunk 6 above)
Comment (ChristiaensBert) [reply src/controlnet_aux/processor.py:125]: I will add SAM in the next PR

Comment (pdoane) [reply src/controlnet_aux/processor.py:125]: I added support for SAM to work with model data that was on HuggingFace in #49, so we can wait for that PR or grab the changes there into a separate one.

**Comment by patrickvonplaten** [comment]
Nice tests! Think we can merge this one

**Pull request merged**